dashboard/app/models/pd/application/teacher_application.rb

```diff
@@ -65,16 +82,263 @@ class Teacher2122Application < TeacherApplicationBase
 
     has_many :emails, class_name: 'Pd::Application::Email', foreign_key: 'pd_application_id'
 
+    validates :status, exclusion: {in: ['interview'], message: '%{value} is reserved for facilitator applications.'}
+    validates :course, presence: true, inclusion: {in: VALID_COURSES}
     validates_uniqueness_of :user_id
     validate :workshop_present_if_required_for_status, if: -> {status_changed?}
 
+    before_validation :set_course_from_program
+    before_save :save_partner, if: -> {form_data_changed? && regional_partner_id.nil? && !deleted?}
     before_save :log_status, if: -> {status_changed?}
 
     serialized_attrs %w(
+      pd_workshop_id
       status_log
       principal_approval_not_required
     )
 
+    # Updates the associated user's school info with the info from this teacher application
+    # based on these rules in order:
+    # 1. Application has a specific school? always overwrite the user's school info
+    # 2. User doesn't have a specific school? overwrite with the custom school info.
+    def update_user_school_info!
+      if school_id || user.school_info.try(&:school).nil?
+        school_info = get_duplicate_school_info(school_info_attr) || SchoolInfo.create!(school_info_attr)
+        user.update_school_info(school_info)
+      end
+    end
+
+    def update_scholarship_status(scholarship_status)
+      Pd::ScholarshipInfo.update_or_create(user, application_year, course, scholarship_status)
+    end
+
+    def scholarship_status
+      if user && application_year && course
+        Pd::ScholarshipInfo.find_by(user: user, application_year: application_year, course: course)&.scholarship_status
+      end
+    end
+
+    # @return a valid year (see ApplicationConstants.APPLICATION_YEARS)
+    def year
+      self.class.year
+    end
+
+    def self.year
+      YEAR_21_22
```

Comment:
For my education, will this line need to be changed when we run the application next year?

It might be helpful to put in a doc somewhere (or a PR?) the set of changes that would need to be made each year.

Reply:
Yeah, I haven't looked into what all the changes are for the next application, but if nothing structurally changes, then changing this constant should be the minimum amount of work